### PR TITLE
More work on conformance / bugfixes

### PR DESCRIPTION
**Pull request by franz**
* memleaks caused by refcount bugs
* images
* external testsuites
* clEnqueue* / clCreate* additional checks

almost everything in test_conformance/basic now passes, exception being some a few  out-of-memory tests and 2 or so which crash in LLVM.

### CODE REVIEW

File: include/_kernel.h [hunk 1]
```diff
@@ -55,6 +55,10 @@
 /* must come after _enable_all_exts.h b/c of pocl_types.h*/
 #include "_kernel_c.h"
 
+#define __kernel_exec(X, typen) \
```
Comment (pjaaskel) [inline include/_kernel.h]: Please add a comment what is the purpose of this macro with two leading underscores.

File: include/_kernel.h [hunk 2]
```diff
@@ -440,14 +442,19 @@ size_t _CL_OVERLOADABLE get_num_groups(uint);
 size_t _CL_OVERLOADABLE get_group_id(uint);
 size_t _CL_OVERLOADABLE get_global_offset(uint);
 
+#if (__OPENCL_C_VERSION__ < 121)
+void read_mem_fence (	cl_mem_fence_flags flags);
```
Comment (pjaaskel) [inline include/_kernel.h]: check the style

File: include/_kernel.h [hunk 3]
```diff
@@ -2395,152 +2445,173 @@ __IF_FP64(_CL_DECLARE_PREFETCH_FUNCS(double));
...
+/* NO sampler */
 
 float4 _CL_OVERLOADABLE read_imagef (IMG_RO_AQ image2d_t image, int2 coord);
-
 float4 _CL_OVERLOADABLE read_imagef (IMG_RO_AQ image2d_array_t image, int4 coord);
+float4 _CL_OVERLOADABLE read_imagef (IMG_RO_AQ image3d_t image, int4 coord);
 
-float4 _CL_OVERLOADABLE read_imagef (IMG_RO_AQ image2d_array_t image, sampler_t sampler,
-                                     int4 coord);
+uint4 _CL_OVERLOADABLE read_imageui ( IMG_RO_AQ image2d_t image, int2 coord);
```
Comment (pjaaskel) [inline include/_kernel.h]: Check the style. Extra whitespace after ( at least. And return value should be in a separate row per GNU C style.

File: lib/CL/clEnqueueReadImage.c [hunk 4]
```diff
@@ -52,9 +52,10 @@ CL_API_SUFFIX__VERSION_1_0
   POCL_RETURN_ERROR_ON((command_queue->context != image->context),
     CL_INVALID_CONTEXT, "image and command_queue are not from the same context\n");
 
-  POCL_RETURN_ERROR_ON((image->buffer->flags & (CL_MEM_HOST_WRITE_ONLY | CL_MEM_HOST_NO_ACCESS)),
-    CL_INVALID_OPERATION, "Image buffer has been created with CL_MEM_HOST_WRITE_ONLY "
-    "or CL_MEM_HOST_NO_ACCESS\n");
+  if (image->buffer)
+    POCL_RETURN_ERROR_ON((image->buffer->flags & (CL_MEM_HOST_WRITE_ONLY | CL_MEM_HOST_NO_ACCESS)),
```
Comment (pjaaskel) [inline lib/CL/clEnqueueReadImage.c]: Long line.

File: lib/CL/clEnqueueWriteImage.c [hunk 5]
```diff
@@ -26,9 +26,10 @@ POname(clEnqueueWriteImage)(cl_command_queue    command_queue,
   POCL_RETURN_ERROR_ON((command_queue->context != image->context),
     CL_INVALID_CONTEXT, "image and command_queue are not from the same context\n");
 
-  POCL_RETURN_ERROR_ON((image->buffer->flags & (CL_MEM_HOST_READ_ONLY | CL_MEM_HOST_NO_ACCESS)),
-    CL_INVALID_OPERATION, "image buffer has been created with CL_MEM_HOST_READ_ONLY "
-    "or CL_MEM_HOST_NO_ACCESS\n");
+  if (image->buffer)
+    POCL_RETURN_ERROR_ON((image->buffer->flags & (CL_MEM_HOST_READ_ONLY | CL_MEM_HOST_NO_ACCESS)),
```
Comment (pjaaskel) [inline lib/CL/clEnqueueWriteImage.c]: Also.

File: include/_kernel.h [hunk 6]
```diff
@@ -444,6 +442,12 @@ size_t _CL_OVERLOADABLE get_num_groups(uint);
 size_t _CL_OVERLOADABLE get_group_id(uint);
 size_t _CL_OVERLOADABLE get_global_offset(uint);
 
+#if (__OPENCL_C_VERSION__ < 121)
+void read_mem_fence (	cl_mem_fence_flags flags);
```
Comment (pjaaskel) [inline include/_kernel.h]: Style.

File: lib/kernel/mem_fence.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+size_t _CL_OVERLOADABLE
+get_local_id(unsigned int dimindx);
+
+void read_mem_fence (	cl_mem_fence_flags flags)
+{
+  asm volatile("" ::: "memory");
```
Comment (pjaaskel) [inline lib/kernel/mem_fence.c]: Did you check if there's an LLVM builtin for the fences which are better used here?

**Comment by franz** [comment]
closing, needs more work

**Pull request closed without merging**